core/bloombits/matcher.go

```diff
@@ -0,0 +1,559 @@
+func (m *Matcher) setTopics(topics [][]common.Hash) {
+	m.topics = nil
+loop:
+	for _, topicList := range topics {
+		t := make([]types.BloomIndexList, len(topicList))
+		for i, b := range topicList {
+			if (b == common.Hash{}) {
+				continue loop
+			}
+			t[i] = types.BloomIndexes(b.Bytes())
+		}
+		m.topics = append(m.topics, t)
+	}
+
+	for _, idxss := range m.topics {
+		for _, idxs := range idxss {
+			for _, idx := range idxs {
+				m.newFetcher(idx)
+			}
+		}
+	}
+}
+
+// match creates a daisy-chain of sub-matchers, one for the address set and one for each topic set, each
+// sub-matcher receiving a section only if the previous ones have all found a potential match in one of
+// the blocks of the section, then binary AND-ing its own matches and forwaring the result to the next one
+func (m *Matcher) match(processCh chan partialMatches, stop chan struct{}) chan partialMatches {
+	subIdx := m.topics
+	if len(m.addresses) > 0 {
+		subIdx = append([][]types.BloomIndexList{m.addresses}, subIdx...)
+	}
+	m.getNextReqCh = make(chan chan nextRequests) // should be a blocking channel
+	m.distributeRequests(stop)
+
+	for _, idx := range subIdx {
+		processCh = m.subMatch(processCh, idx, stop)
+	}
+	return processCh
+}
+
+// partialMatches with a non-nil vector represents a section in which some sub-matchers have already
+// found potential matches. Subsequent sub-matchers will binary AND their matches with this vector.
+// If vector is nil, it represents a section to be processed by the first sub-matcher.
+type partialMatches struct {
+	sectionIdx uint64
+	vector     []byte
+}
+
+// newFetcher adds a fetcher for the given bit index if it has not existed before
+func (m *Matcher) newFetcher(idx uint) {
+	if _, ok := m.fetchers[idx]; ok {
+		return
+	}
+	f := &fetcher{
+		bitIdx: idx,
+		reqMap: make(map[uint64]req),
+	}
+	m.fetchers[idx] = f
+}
+
+// subMatch creates a sub-matcher that filters for a set of addresses or topics, binary OR-s those matches, then
+// binary AND-s the result to the daisy-chain input (processCh) and forwards it to the daisy-chain output.
+// The matches of each address/topic are calculated by fetching the given sections of the three bloom bit indexes belonging to
+// that address/topic, and binary AND-ing those vectors together.
+func (m *Matcher) subMatch(processCh chan partialMatches, idxs []types.BloomIndexList, stop chan struct{}) chan partialMatches {
+	// set up fetchers
+	fetchIdx := make([][3]chan uint64, len(idxs))
+	fetchData := make([][3]chan []byte, len(idxs))
+	for i, idx := range idxs {
+		for j, ii := range idx {
+			fetchIdx[i][j] = make(chan uint64, channelCap)
+			fetchData[i][j] = m.fetchers[ii].fetch(fetchIdx[i][j], m.distCh, stop, &m.wg)
+		}
+	}
+
+	fetchedCh := make(chan partialMatches, channelCap) // entries from processCh are forwarded here after fetches have been initiated
+	resultsCh := make(chan partialMatches, channelCap)
+
+	m.wg.Add(2)
+	// goroutine for starting retrievals
+	go func() {
```

Comment:
It's unclear to me why there need to be two independent goroutines here, or how they work together. Can you elaborate?